### ISSUE TITLE
Tile palette bug thread

### ISSUE THREAD

**Issue by Raptorizer** [unitystation/unitystation#3280]
## Description
Below is all the things that is wrong with the Tile Palette and needs to be fixed ASAP.

- Disappearing tiles within tile palette via dropdown menu
This occurs in every single tile palette with some not being as bad as others. The categories marked in red don't show up at all. 
![6D6VMr5kdY](https://user-images.githubusercontent.com/44679539/75091726-1eb51400-5536-11ea-9d1e-fcac6e149ec7.gif)

- The tool 'Generate Tile Palettes' does nothing
![NdMKTGgK48](https://user-images.githubusercontent.com/44679539/75091641-4a83ca00-5535-11ea-904f-46743f31444c.png)
- Selecting Palettes via the asset folder displays the icons but then immediately disappears upon mouse hovering within the window
![AVyXDnaEIY](https://user-images.githubusercontent.com/44679539/75091700-fd542800-5535-11ea-8a07-bc8ef82f4b9f.gif)
- Tiles are associated incorrectly
![LUDDEmgFkT](https://user-images.githubusercontent.com/44679539/75091786-c6cadd00-5536-11ea-9304-ceca8302896a.gif)

If more issues are found, comment below so they can be added

**Comment by ThatDan123** [comment]
If you change this line the tile palette becomes more usable: 
![image](https://user-images.githubusercontent.com/56727168/75092017-90f41b80-556b-11ea-8b90-a58c7dcf27df.png)


![image](https://user-images.githubusercontent.com/56727168/75092021-9ea9a100-556b-11ea-9a38-ac53200a2ce8.png)

Just uncomment the first bit, (already uncommented in image)

**Comment by Raptorizer** [comment]
If that is the case, can that change be made and submitted as a PR? Or is it local only?

**Comment by ThatDan123** [comment]
I wouldnt PR as im not sure if it would affect performance and also it doesnt fix all the issues so theres still an underlying problem somewhere and this might hide the bug.

**Comment by DVanwart** [comment]
I tried digging around with this, so when everything disappears it's (I think) due to Unity calling GetTileData on every tile, while the PreviewSprite is null.
```
public abstract class GenericTile : TileBase
{
	public virtual Sprite PreviewSprite { get; set; }

	public override void GetTileData(Vector3Int position, ITilemap tilemap, ref TileData tileData)
	{
		tileData.sprite = PreviewSprite;
		tileData.flags = TileFlags.None;
		tileData.colliderType = Tile.ColliderType.Grid;
	}
}
```

I found adding this to ObjectTile fixes part of the issue (OnValidate doesn't allways run before issue triggers so it didn't always fix it). LoadSprite() has no real performace impact unlike Create().
```
private void OnEnable()
{
	PreviewSprite = PreviewSpriteBuilder.LoadSprite(Object);
}
```

**Comment by DVanwart** [comment]
That said there is still a secondary issue where the tiles that appear don't fully match the initial tile palette. 
Before mouseover
![image](https://user-images.githubusercontent.com/17316823/75716123-fd29fa00-5ca4-11ea-8e21-d58dae8aca97.png)

After mouseover
![image](https://user-images.githubusercontent.com/17316823/75716160-07e48f00-5ca5-11ea-8e4a-b533b17b2fd5.png)

This includes the weird pink tiles appearing on the Containers and other palettes.

**Comment by DVanwart** [comment]
So I've run into a few walls!

- PreviewSpriteBuilder.Create() is needed when an object doesn't have a preview sprite already, else they just end up blank. But if .Create() Fails it just keep running over and over freezing Unity.
- I have no idea where the pink tiles are coming from, it seems to happen after all the other issues.

I'll post here if I figure this out but otherwise assume I'm stuck, feel free to go digging!

**Comment by lwilson2048** [comment]
I applied the change ThatDan123 mentioned and then added a log message to `PreviewSpriteBuilder.LoadSprite` to see why the preview sprites weren't loading. It appears the issue is that the image files have been moved or deleted. 

For instance, Unity is looking for a sprite at 
`"Assets/Textures/TilePreviews/res/Prefabs/Objects/Shuttle/Shuttle_engine_W.png"`
which has been moved to 
`"Assets/Textures/TilePreviews/res/prefabs/Objects/Static_Placeholders/Shuttle/"`.

I tested simply moving "Shuttle_engine_W.png" to where Unity expected to find it which fixed the issue for that sprite.

I'm a newbie when it comes to Unity and this project so I'm not sure what the easiest way to fix this would be. 

I have attached a text file of all the moved or missing files.
[missing_or_moved.txt](https://github.com/unitystation/unitystation/files/4300044/missing_or_moved.txt)
